### PR TITLE
Bump controller-runtime, k8s libraries, openshift/library go, adjust CO to the new versions

### PR DESCRIPTION
**Pull request by jhrozek**
- fix(deps): update module sigs.k8s.io/controller-runtime to v0.15.0
- Bump openshift/library-go
- bump k8s.io/pod-security-admission to v0.27.2
- bump sigs.k8s.io/controller-tools to v0.12.0
- Adjust CO reconcile loop setup and unit tests for the new controller-manager version

Adjust CO reconcile loop setup and unit tests for the new controller-manager version

In order to use the latest release of controller-manager we need to use
the new style of setting up controller loops that uses
ctrl.NewControllerManagedBy() and then the builder pattern instead of
adding c.Watch for each resource.

The tests had to be adjusted, too, to specifically name resources that
can update its status.

### CODE REVIEW

**Comment by jhrozek** [comment]
hmm, that test passes locally when I run that test only. But the last three CI runs of this PR fail always on that single test. Running more tests locally.

**Comment by jhrozek** [comment]
hmm all the parallel tests are passing locally. I'm going to re-run the tests again and watch them interactively to try and catch the error..

**Comment by jhrozek** [comment]
I /think/ I managed to catch the issue:
```
{"level":"error","ts":"2023-06-28T14:38:36.312Z","logger":"remediationctrl","msg":"Retriable error","Request.Namespace":"osdk-e2e-ac518a1e-a339-4401-9a13-72e5
b2162618","Request.Name":"test-generic-remediation-fails-unkown","error":"failed to get API group resources: unable to retrieve the complete list of server APIs: foo.bar/v1: the server could not find the requested resource","stacktrace":"github.com/ComplianceAsCode/compliance-operator/pkg/controller/common.ReturnWi
thRetriableError\n\tgithub.com/ComplianceAsCode/compliance-operator/pkg/controller/common/errors.go:117\ngithub.com/ComplianceAsCode/compliance-operator/pkg/controller/complianceremediation.(*ReconcileComplianceRemediation).Reconcile\n\tgithub.com/ComplianceAsCode/compliance-operator/pkg/controller/complianceremedi
ation/complianceremediation_controller.go:173\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/cont
roller-runtime@v0.15.0/pkg/internal/controller/controller.go:314\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.f
unc2.2\n\tsigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226"}                                                                  {"level":"error","ts":"2023-06-28T14:38:36.312Z","msg":"Reconciler error","controller":"complianceremediation-controller","controllerGroup":"compliance.opensh
ift.io","controllerKind":"ComplianceRemediation","ComplianceRemediation":{"name":"test-generic-remediation-fails-unkown","namespace":"osdk-e2e-ac518a1e-a339-4401-9a13-72e5b2162618"},"namespace":"osdk-e2e-ac518a1e-a339-4401-9a13-72e5b2162618","name":"test-generic-remediation-fails-unkown","reconcileID":"7c2cba67-3cd
4-4d64-9fd2-fa02cc484099","error":"failed to get API group resources: unable to retrieve the complete list of server APIs: foo.bar/v1: the server could not find the requested resource","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runti
me@v0.15.0/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsi
gs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226"}                                                                 
```
I guess the problem is that this should not be a `Retriable` Error..

**Comment by jhrozek** [comment]
/hold
because the commit I added is a dirty hack and we should do better. But I want to see how the CI fares now (manual tests passed with the workaround)

File: pkg/controller/complianceremediation/complianceremediation_controller.go [hunk 1]
```diff
@@ -219,7 +219,7 @@ func (r *ReconcileComplianceRemediation) reconcileRemediation(instance *compv1al
 		return common.NewNonRetriableCtrlError(
 			"Unable to get fix object from ComplianceRemediation. "+
 				"Please update the compliance-operator's permissions: %w", err)
-	} else if runtime.IsNotRegisteredError(err) || meta.IsNoMatchError(err) {
+	} else if runtime.IsNotRegisteredError(err) || meta.IsNoMatchError(err) || (err != nil && strings.Contains(err.Error(), "the server could not find the requested resource")) {
```
Comment (jhrozek) [inline pkg/controller/complianceremediation/complianceremediation_controller.go:222]: OK, so this is the ugly workaround. I cannot for the life of me figure out how to catch the proper error. In the error string chain I see `unable to retrieve the complete list of server APIs` which would suggest the error is `ErrGroupDiscoveryFailed` which should be testable for with `IsGroupDiscoveryFailedError` from `k8s.io/client-go/discovery/discovery_client.go` but that doesn't seem to work for me. Which is especially odd with go 1.20 where `errors.Is` should even work across wrapped errors. I'll continue trying to get to the root of the issue, but as the worst case solution, maybe this workaround would at least pass CI.

Comment (rhmdnd) [reply pkg/controller/complianceremediation/complianceremediation_controller.go:222]: Looks like it worked for CI at least.

**Comment by jhrozek** [comment]
I included the patch from PR #370 as a separate patch because go mod tidy was also needed and squashing would be too messy.

File: pkg/controller/complianceremediation/complianceremediation_controller.go (hunk 1 above)
Comment (rhmdnd) [reply pkg/controller/complianceremediation/complianceremediation_controller.go:222]: Strange - after looking at the implementation in `client-go` I'm not sure why we wouldn't be able to use:

```go
IsGroupDiscoveryFailedError(err)
```

**Comment by rhmdnd** [comment]
I'm fine merging this for now and cleaning up the error after so we can at least get updated dependencies. But, yeah, that is odd...

**Comment by rhmdnd** [comment]
Adding other appropriate labels since this is a dep bump.

**Comment by rhmdnd** [comment]
Making sure @Vincent056 and @yuumasato have an opportunity to take a look at this before we remove the hold.

**Comment by yuumasato** [comment]
> Making sure @Vincent056 and @yuumasato have an opportunity to take a look at this before we remove the hold.

@rhmdnd Thanks. LGTM, based on my little knowledge of Go and the operator.

**Comment by rhmdnd** [comment]
Getting this merged so we can cleanup the dep bump patches.

**Pull request merged**